### ISSUE TITLE
Split code into modules

### ISSUE THREAD

**Issue by Morganamilo** [Jguer/yay#759]
Ideally I want yay to be built out of small, "do one thing and do it well" blocks. go-alpm, aur, go-srcinfo, go-pacmanconf are all good examples of this.

I'd like to split yay into more pieces but do not really see anything that should be extracted to a new project. Instead I think it would be a good idea to split this repo into many modules, similar to how go-pacmanconf has its own ini module.

I believe yay was like this a very long time ago, before I started contributing. Any reason you moved to one main module? And would you be apposed to splitting it out again?

Currently the two modules I want to create is one for all the coloured output. And a wrapper around aur that multi threads and respects requestsplitn.

**Comment by Jguer** [comment]
I used to have yay split into `aur`, `print`, `repo` , something along those lines, but `rpc` was delegated to Mikel Oscars AUR, so it lost our `aur` lost all re-usability since it started to contain only yay specific code, that and the interactions between aur and repo made me scrap the modules. Not that the modules were bad, they were just badly divided because yay grew organically and there was no interest in having a public API for them. 

If I were to split it out again I would definitively make a config module and a devel module. 
I would be in favor of it but it has to be well thought out to avoid having to break it apart later again because modules require too much knowledge of each other

**Comment by Morganamilo** [comment]
>and finally get that vendor directory out of the repo

Well the way I setup the pr right now is that it still uses a vendor directory, otherwise it would pollute home or tmp or whatever, which is not good for packaging. We can remove vendor/ from the repo and download it at build time. But then that could be done just as easily with dep.

**Comment by Morganamilo** [comment]
This goes back to #579 a little. Ideally I'd like to create a local repo based helper and I don't want to duplicate code. Even ignoring that, having code separation improves readability and maintainability.

I will go and split the code into modules then (probably after #705 is merged). I'll split just the bits that I think would make good self contained modules. I'm not going to try and split the entire code base just cause.

**Comment by alxpettit** [comment]
If it's split, I would very much like to have it retain a single command for the interface. Otherwise I imagine it'll end up being like apt-get/dpkg with lots of functionality scattered around and poor discoverability... so, definitely want to avoid that.

**Comment by Jguer** [comment]
The module split would not affect current UX, instead it would just make it easier to navigate through the code for developers and compartmentalizing functionality to specific modules normally makes it easier to swap or make changes to components because of a stable application interface. 

A bad module architecture can be self defeating in that you end up having to hack components together or tying them too close therefor annulling all benefits.

I've thinking about this and everything that all of the dep logic/package resolution should be modularized as @Morganamilo also said, could be useful for other projects.

**Comment by alxpettit** [comment]
So library-level modularity. I'm definitely for that!
